packages/runner/src/storage/cache.ts

```diff
@@ -1379,10 +1444,767 @@ export const getChanges = <
+  requested: MemorySpace;
+  constructor(
+    { open, requested }: { open: MemorySpace; requested: MemorySpace },
+  ) {
+    super(
+      `Can not open transaction writer for ${requested} beacuse transaction has writer open for ${open}`,
+    );
+    this.open = open;
+    this.requested = requested;
+  }
+}
+
+export class NotFound extends RangeError implements INotFoundError {
+  override name = "NotFoundError" as const;
+}
+
+function addressToKey(address: IMemoryAddress): string {
+  return `${address.id}/${address.type}/${address.path.join("/")}`;
+}
+
+/**
+ * Convert IMemoryAddress to FactAddress for use with existing storage system.
+ */
+function toFactAddress(address: IMemoryAddress): FactAddress {
+  return {
+    the: address.type,
+    of: address.id,
+  };
+}
+
+/**
+ * Reads the value from the given fact at a given path and either returns
+ * {@link Read} object or {@link NotFoundError} if path does not exist in
+ * the provided {@link State}.
+ *
+ * Read fails when key is accessed in the non-existing parent, but succeeds
+ * with `undefined` when last component of the path does not exists. Here are
+ * couple of examples illustrating behavior.
+ *
+ * ```ts
+ * const unclaimed = {
+ *    the: "application/json",
+ *    of: "test:1",
+ * }
+ * const fact = {
+ *    the: "application/json",
+ *    of: "test:1",
+ *    is: { hello: "world", from: { user: { name: "Alice" } } }
+ * }
+ *
+ * read({ path: [] }, fact)                   // { ok: { value: fact.is } }
+ * read({ path: ['hello'] }, fact)            // { ok: { value: "world" } }
+ * read({ path: ['hello', 'length'] }, fact)  // { ok: { value: undefined } }
+ * read({ path: ['hello', 0] }, fact)         // { ok: { value: undefined } }
+ * read({ path: ['hello', 0, 0] }, fact)      // { error }
+ * read({ path: ['from', 'user'] }, fact)     // { ok: { value: {name: "Alice"} } }
+ * read({ path: [] }, unclaimed)              // { ok: { value: undefined } }
+ * read({ path: ['a'] }, unclaimed)           // { error }
+ * ```
+ */
+const read = (
+  state: Assert | Retract,
+  address: IMemoryAddress,
+): Result<{ value?: JSONValue }, INotFoundError> => resolve(state, address);
+
+const resolve = (
+  state: Assert | Retract,
+  address: IMemoryAddress,
+) => {
+  const { path } = address;
+  let value = state?.is as JSONValue | undefined;
+  let at = -1;
+  while (++at < path.length) {
+    const key = path[at];
+    if (typeof value === "object" && value != null) {
+      // We do not support array.length as that is JS specific getter.
+      value = Array.isArray(value) && key === "length"
+        ? undefined
+        : (value as Record<string, JSONValue>)[key];
+    } else {
+      return {
+        error: state
+          ? new NotFound(
+            `Can not resolve "${address.type}" of "${address.id}" at "${
+              path.slice(0, at).join(".")
+            }" in "${address.space}", because target is not an object`,
+          )
+          : new NotFound(
```

Comment:
should this be a different error type?

or maybe add a field to the error that has an optional `path` that if the document did exist, it shows what part of the path matched (so the reader can use that as hint from where on the construct things)